robot-server/robot_server/service/task_runner.py

```diff
@@ -36,15 +34,45 @@ def run(self, func: TaskFunc, **kwargs: Any) -> None:
 
         Arguments:
             func: An async, None-returning function to run in the background.
-            Use kwargs to add arguments if required.
+            Use **kwargs to pass to func.
         """
         func_name = func.__qualname__
 
-        async def _run_async_task() -> None:
+        async def wrapper() -> None:
+            # Add the previous logic back
+            current_task = asyncio.current_task()
             try:
                 await func(**kwargs)
                 log.debug(f"Background task {func_name} succeeded")
             except Exception as e:
                 log.warning(f"Background task {func_name} failed", exc_info=e)
+            finally:
+                self._running_tasks.remove(current_task)
+
+        wrapper_task = asyncio.create_task(wrapper())
+        self._running_tasks.add(wrapper_task)
+
+    async def cancel_all_and_clean_up(self) -> None:
+        for task in self._running_tasks:
+            task.cancel()
+        await asyncio.gather(*self._running_tasks, return_exceptions=True)
+
+
+def initialize_task_runner(app_state: AppState) -> None:
+    """Create a new `TaskRunner` and store it on `app_state`."""
+    _init_taskrunner.set_on(app_state, TaskRunner())
+
+
+async def clean_up_task_runner(app_state: AppState) -> None:
+    """Clean up the `TaskRunner` stored on `app_state`."""
+    task_runner = _init_taskrunner.get_from(app_state)
+
+    if task_runner is not None:
+        await task_runner.cancel_all_and_clean_up()
+
 
-        self._background_tasks.add_task(_run_async_task)
+# The get_task_runner method is for the HTTP endpoint functions that need a TaskRunner instance.
+def get_task_runner(app_state: AppState = Depends(get_app_state)) -> TaskRunner:
```

Comment:
Let's turn this comment into a docstring.

```suggestion
def get_task_runner(app_state: AppState = Depends(get_app_state)) -> TaskRunner:
    """Return the server's `TaskRunner`.
    
    Intended to be used by endpoint functions as a FastAPI dependency,
    like `task_runner = fastapi.Depends(get_task_runner)`.
    """
```